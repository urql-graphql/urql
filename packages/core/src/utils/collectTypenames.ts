interface EntityLike {
  [key: string]: EntityLike | EntityLike[] | any;
  __typename: string | null | void;
}

const collectTypes = (obj: EntityLike | EntityLike[], types: Set<string>) => {
  if (Array.isArray(obj)) {
    for (const item of obj) collectTypes(item, types);
  } else if (typeof obj === 'object' && obj !== null) {
    for (const key in obj) {
      if (key === '__typename' && typeof obj[key] === 'string') {
        types.add(obj[key] as string);
      } else {
        collectTypes(obj[key], types);
      }
    }
  }

  return types;
};

/** Finds and returns a list of `__typename` fields found in response data.
 *
 * @privateRemarks
 * This is used by `@urql/core`’s document `cacheExchange` to find typenames
 * in a given GraphQL response’s data.
 */
export const collectTypenames = (response: object): string[] => [
  ...collectTypes(response as EntityLike, new Set()),
];
