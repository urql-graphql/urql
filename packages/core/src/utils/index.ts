export * from './error';
export * from './request';
export * from './result';
export * from './variables';
export * from './collectTypenames';
export * from './formatDocument';
export * from './maskTypename';
export * from './streamUtils';
export * from './operation';

export const noop = () => {
  /* noop */
};
