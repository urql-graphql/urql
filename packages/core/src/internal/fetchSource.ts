import { Source, make } from 'wonka';
import { Operation, OperationResult } from '../types';
import { makeResult, makeErrorResult, mergeResultPatch } from '../utils';

const asyncIterator =
  typeof Symbol !== 'undefined' ? Symbol.asyncIterator : null;
const decoder = typeof TextDecoder !== 'undefined' ? new TextDecoder() : null;
const jsonHeaderRe = /content-type:[^\r\n]*application\/json/i;
const boundaryHeaderRe = /boundary="?([^=";]+)"?/i;

type ChunkData = { done: false; value: Buffer | Uint8Array } | { done: true };

// NOTE: We're avoiding referencing the `Buffer` global here to prevent
// auto-polyfilling in Webpack
const toString = (input: Buffer | ArrayBuffer): string =>
  input.constructor.name === 'Buffer'
    ? (input as Buffer).toString()
    : decoder!.decode(input as ArrayBuffer);

// DERIVATIVE: Copyright (c) 2021 Marais Rossouw <hi@marais.io>
// See: https://github.com/maraisr/meros/blob/219fe95/src/browser.ts
const executeIncrementalFetch = (
  onResult: (result: OperationResult) => void,
  operation: Operation,
  response: Response
): Promise<void> => {
  // NOTE: Guarding against fetch polyfills here
  const contentType =
    (response.headers && response.headers.get('Content-Type')) || '';
  if (!/multipart\/mixed/i.test(contentType)) {
    return response.json().then(payload => {
      onResult(makeResult(operation, payload, response));
    });
  }

  let boundary = '---';
  const boundaryHeader = contentType.match(boundaryHeaderRe);
  if (boundaryHeader) boundary = '--' + boundaryHeader[1];

  let read: () => Promise<ChunkData>;
  let cancel = () => {
    /*noop*/
  };
  if (asyncIterator && response[asyncIterator]) {
    const iterator = response[asyncIterator]();
    read = iterator.next.bind(iterator);
  } else if ('body' in response && response.body) {
    const reader = response.body.getReader();
    cancel = reader.cancel.bind(reader);
    read = reader.read.bind(reader);
  } else {
    throw new TypeError('Streaming requests unsupported');
  }

  let buffer = '';
  let isPreamble = true;
  let nextResult: OperationResult | null = null;
  let prevResult: OperationResult | null = null;

  function next(data: ChunkData): Promise<void> | void {
    if (!data.done) {
      const chunk = toString(data.value);
      let boundaryIndex = chunk.indexOf(boundary);
      if (boundaryIndex > -1) {
        boundaryIndex += buffer.length;
      } else {
        boundaryIndex = buffer.indexOf(boundary);
      }

      buffer += chunk;
      while (boundaryIndex > -1) {
        const current = buffer.slice(0, boundaryIndex);
        const next = buffer.slice(boundaryIndex + boundary.length);

        if (isPreamble) {
          isPreamble = false;
        } else {
          const headersEnd = current.indexOf('\r\n\r\n') + 4;
          const headers = current.slice(0, headersEnd);
          const body = current.slice(headersEnd, current.lastIndexOf('\r\n'));

          let payload: any;
          if (jsonHeaderRe.test(headers)) {
            try {
              payload = JSON.parse(body);
              nextResult = prevResult = prevResult
                ? mergeResultPatch(prevResult, payload, response)
                : makeResult(operation, payload, response);
            } catch (_error) {}
          }

          if (next.slice(0, 2) === '--' || (payload && !payload.hasNext)) {
            if (!prevResult)
              return onResult(makeResult(operation, {}, response));
            break;
          }
        }

        buffer = next;
        boundaryIndex = buffer.indexOf(boundary);
      }
    }

    if (nextResult) {
      onResult(nextResult);
      nextResult = null;
    }

    if (!data.done && (!prevResult || prevResult.hasNext)) {
      return read().then(next);
    }
  }

  return read().then(next).finally(cancel);
};

export const makeFetchSource = (
  operation: Operation,
  url: string,
  fetchOptions: RequestInit
): Source<OperationResult> => {
  const maxStatus = fetchOptions.redirect === 'manual' ? 400 : 300;
  const fetcher = operation.context.fetch;

  return make<OperationResult>(({ next, complete }) => {
    const abortController =
      typeof AbortController !== 'undefined' ? new AbortController() : null;
    if (abortController) {
      fetchOptions.signal = abortController.signal;
    }

    let ended = false;
    let statusNotOk = false;
    let response: Response;

    Promise.resolve()
      .then(() => {
        if (ended) return;
        return (fetcher || fetch)(url, fetchOptions);
      })
      .then((_response: Response | void) => {
        if (!_response) return;
        response = _response;
        statusNotOk = response.status < 200 || response.status >= maxStatus;
        return executeIncrementalFetch(next, operation, response);
      })
      .then(complete)
      .catch((error: Error) => {
        if (error.name !== 'AbortError') {
          const result = makeErrorResult(
            operation,
            statusNotOk ? new Error(response.statusText) : error,
            response
          );

          next(result);
          complete();
        }
      });

    return () => {
      ended = true;
      if (abortController) {
        abortController.abort();
      }
    };
  });
};
