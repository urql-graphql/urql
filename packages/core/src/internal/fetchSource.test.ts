import { pipe, scan, subscribe, toPromise } from 'wonka';

import { queryOperation, context } from '../test-utils';
import { makeFetchSource } from './fetchSource';
import { gql } from '../gql';
import { OperationResult, Operation } from '../types';
import { makeOperation } from '../utils';

const fetch = (global as any).fetch as jest.Mock;
const abort = jest.fn();

const abortError = new Error();
abortError.name = 'AbortError';

beforeAll(() => {
  (global as any).AbortController = function AbortController() {
    this.signal = undefined;
    this.abort = abort;
  };
});

beforeEach(() => {
  fetch.mockClear();
  abort.mockClear();
});

afterAll(() => {
  (global as any).AbortController = undefined;
});

const response = {
  status: 200,
  data: {
    data: {
      user: 1200,
    },
  },
};

describe('on success', () => {
  beforeEach(() => {
    fetch.mockResolvedValue({
      status: 200,
      json: jest.fn().mockResolvedValue(response),
    });
  });

  it('returns response data', async () => {
    const fetchOptions = {};
    const data = await pipe(
      makeFetchSource(queryOperation, 'https://test.com/graphql', fetchOptions),
      toPromise
    );

    expect(data).toMatchSnapshot();

    expect(fetch).toHaveBeenCalled();
    expect(fetch.mock.calls[0][0]).toBe('https://test.com/graphql');
    expect(fetch.mock.calls[0][1]).toBe(fetchOptions);
  });

  it('uses the mock fetch if given', async () => {
    const fetchOptions = {};
    const fetcher = jest.fn().mockResolvedValue({
      status: 200,
      json: jest.fn().mockResolvedValue(response),
    });

    const data = await pipe(
      makeFetchSource(
        {
          ...queryOperation,
          context: {
            ...queryOperation.context,
            fetch: fetcher,
          },
        },
        'https://test.com/graphql',
        fetchOptions
      ),
      toPromise
    );

    expect(data).toMatchSnapshot();
    expect(fetch).not.toHaveBeenCalled();
    expect(fetcher).toHaveBeenCalled();
  });
});

describe('on error', () => {
  beforeEach(() => {
    fetch.mockResolvedValue({
      status: 400,
      json: jest.fn().mockResolvedValue({}),
    });
  });

  it('returns error data', async () => {
    const fetchOptions = {};
    const data = await pipe(
      makeFetchSource(queryOperation, 'https://test.com/graphql', fetchOptions),
      toPromise
    );

    expect(data).toMatchSnapshot();
  });

  it('returns error data with status 400 and manual redirect mode', async () => {
    const data = await pipe(
      makeFetchSource(queryOperation, 'https://test.com/graphql', {
        redirect: 'manual',
      }),
      toPromise
    );

    expect(data).toMatchSnapshot();
  });

  it('ignores the error when a result is available', async () => {
    const data = await pipe(
      makeFetchSource(queryOperation, 'https://test.com/graphql', {}),
      toPromise
    );

    expect(data).toMatchSnapshot();
  });
});

describe('on teardown', () => {
  it('does not start the outgoing request on immediate teardowns', () => {
    fetch.mockRejectedValue(abortError);

    const { unsubscribe } = pipe(
      makeFetchSource(queryOperation, 'https://test.com/graphql', {}),
      subscribe(fail)
    );

    unsubscribe();
    expect(fetch).toHaveBeenCalledTimes(0);
    expect(abort).toHaveBeenCalledTimes(1);
  });

  it('aborts the outgoing request', async () => {
    fetch.mockRejectedValue(abortError);

    const { unsubscribe } = pipe(
      makeFetchSource(queryOperation, 'https://test.com/graphql', {}),
      subscribe(fail)
    );

    await Promise.resolve();

    unsubscribe();
    expect(fetch).toHaveBeenCalledTimes(1);
    expect(abort).toHaveBeenCalledTimes(1);
  });
});

describe('on multipart/mixed', () => {
  const wrap = (json: object) =>
    '\r\n' +
    'Content-Type: application/json; charset=utf-8\r\n\r\n' +
    JSON.stringify(json) +
    '\r\n---';

  it('listens for more responses (stream)', async () => {
    fetch.mockResolvedValue({
      status: 200,
      headers: {
        get() {
          return 'multipart/mixed';
        },
      },
      body: {
        getReader: function () {
          let cancelled = false;
          const results = [
            {
              done: false,
              value: Buffer.from('\r\n---'),
            },
            {
              done: false,
              value: Buffer.from(
                wrap({
                  hasNext: true,
                  data: {
                    author: {
                      id: '1',
                      name: 'Steve',
                      __typename: 'Author',
                      todos: [{ id: '1', text: 'stream', __typename: 'Todo' }],
                    },
                  },
                })
              ),
            },
            {
              done: false,
              value: Buffer.from(
                wrap({
                  path: ['author', 'todos', 1],
                  data: { id: '2', text: 'defer', __typename: 'Todo' },
                  hasNext: true,
                })
              ),
            },
            {
              done: false,
              value: Buffer.from(wrap({ hasNext: false }) + '--'),
            },
            { done: true },
          ];
          let count = 0;
          return {
            cancel: function () {
              cancelled = true;
            },
            read: function () {
              if (cancelled) throw new Error('No');

              return Promise.resolve(results[count++]);
            },
          };
        },
      },
    });

    const streamedQueryOperation: Operation = makeOperation(
      'query',
      {
        query: gql`
          query {
            author {
              id
              name
              todos @stream {
                id
                text
              }
            }
          }
        `,
        variables: {},
        key: 1,
      },
      context
    );

    const chunks: OperationResult[] = await pipe(
      makeFetchSource(streamedQueryOperation, 'https://test.com/graphql', {}),
      scan((prev: OperationResult[], item) => [...prev, item], []),
      toPromise
    );

    expect(chunks.length).toEqual(3);

    expect(chunks[0].data).toEqual({
      author: {
        id: '1',
        name: 'Steve',
        __typename: 'Author',
        todos: [{ id: '1', text: 'stream', __typename: 'Todo' }],
      },
    });

    expect(chunks[1].data).toEqual({
      author: {
        id: '1',
        name: 'Steve',
        __typename: 'Author',
        todos: [
          { id: '1', text: 'stream', __typename: 'Todo' },
          { id: '2', text: 'defer', __typename: 'Todo' },
        ],
      },
    });

    expect(chunks[2].data).toEqual({
      author: {
        id: '1',
        name: 'Steve',
        __typename: 'Author',
        todos: [
          { id: '1', text: 'stream', __typename: 'Todo' },
          { id: '2', text: 'defer', __typename: 'Todo' },
        ],
      },
    });
  });

  it('listens for more responses (defer)', async () => {
    fetch.mockResolvedValue({
      status: 200,
      headers: {
        get() {
          return 'multipart/mixed';
        },
      },
      body: {
        getReader: function () {
          let cancelled = false;
          const results = [
            {
              done: false,
              value: Buffer.from('\r\n---'),
            },
            {
              done: false,
              value: Buffer.from(
                wrap({
                  hasNext: true,
                  data: {
                    author: {
                      id: '1',
                      __typename: 'Author',
                    },
                  },
                })
              ),
            },
            {
              done: false,
              value: Buffer.from(
                wrap({
                  path: ['author'],
                  data: { name: 'Steve' },
                  hasNext: true,
                })
              ),
            },
            {
              done: false,
              value: Buffer.from(wrap({ hasNext: false }) + '--'),
            },
            { done: true },
          ];
          let count = 0;
          return {
            cancel: function () {
              cancelled = true;
            },
            read: function () {
              if (cancelled) throw new Error('No');

              return Promise.resolve(results[count++]);
            },
          };
        },
      },
    });

    const AuthorFragment = gql`
      fragment authorFields on Author {
        name
      }
    `;

    const streamedQueryOperation: Operation = makeOperation(
      'query',
      {
        query: gql`
          query {
            author {
              id
              ...authorFields @defer
            }
          }

          ${AuthorFragment}
        `,
        variables: {},
        key: 1,
      },
      context
    );

    const chunks: OperationResult[] = await pipe(
      makeFetchSource(streamedQueryOperation, 'https://test.com/graphql', {}),
      scan((prev: OperationResult[], item) => [...prev, item], []),
      toPromise
    );

    expect(chunks.length).toEqual(3);

    expect(chunks[0].data).toEqual({
      author: {
        id: '1',
        __typename: 'Author',
      },
    });

    expect(chunks[1].data).toEqual({
      author: {
        id: '1',
        name: 'Steve',
        __typename: 'Author',
      },
    });

    expect(chunks[2].data).toEqual({
      author: {
        id: '1',
        name: 'Steve',
        __typename: 'Author',
      },
    });
  });

  it('listens for more responses (defer-neted)', async () => {
    fetch.mockResolvedValue({
      status: 200,
      headers: {
        get() {
          return 'multipart/mixed';
        },
      },
      body: {
        getReader: function () {
          let cancelled = false;
          const results = [
            {
              done: false,
              value: Buffer.from('\r\n---'),
            },
            {
              done: false,
              value: Buffer.from(
                wrap({
                  hasNext: true,
                  data: {
                    author: {
                      id: '1',
                      name: 'Steve',
                      address: {
                        country: 'UK',
                        __typename: 'Address',
                      },
                      __typename: 'Author',
                    },
                  },
                })
              ),
            },
            {
              done: false,
              value: Buffer.from(
                wrap({
                  path: ['author', 'address'],
                  data: { street: 'home' },
                  hasNext: true,
                })
              ),
            },
            {
              done: false,
              value: Buffer.from(wrap({ hasNext: false }) + '--'),
            },
            { done: true },
          ];
          let count = 0;
          return {
            cancel: function () {
              cancelled = true;
            },
            read: function () {
              if (cancelled) throw new Error('No');

              return Promise.resolve(results[count++]);
            },
          };
        },
      },
    });

    const AddressFragment = gql`
      fragment addressFields on Address {
        street
      }
    `;

    const streamedQueryOperation: Operation = makeOperation(
      'query',
      {
        query: gql`
          query {
            author {
              id
              address {
                id
                country
                ...addressFields @defer
              }
            }
          }

          ${AddressFragment}
        `,
        variables: {},
        key: 1,
      },
      context
    );

    const chunks: OperationResult[] = await pipe(
      makeFetchSource(streamedQueryOperation, 'https://test.com/graphql', {}),
      scan((prev: OperationResult[], item) => [...prev, item], []),
      toPromise
    );

    expect(chunks.length).toEqual(3);

    expect(chunks[0].data).toEqual({
      author: {
        id: '1',
        name: 'Steve',
        address: {
          country: 'UK',
          __typename: 'Address',
        },
        __typename: 'Author',
      },
    });

    expect(chunks[1].data).toEqual({
      author: {
        id: '1',
        name: 'Steve',
        address: {
          country: 'UK',
          street: 'home',
          __typename: 'Address',
        },
        __typename: 'Author',
      },
    });
  });
});
