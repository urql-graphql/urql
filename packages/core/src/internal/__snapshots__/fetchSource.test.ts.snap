// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`on error ignores the error when a result is available 1`] = `
Object {
  "data": undefined,
  "error": [CombinedError: [Network] ],
  "extensions": undefined,
  "operation": Object {
    "context": Object {
      "fetchOptions": Object {
        "method": "POST",
      },
      "requestPolicy": "cache-first",
      "url": "http://localhost:3000/graphql",
    },
    "key": 2,
    "kind": "query",
    "query": Object {
      "__key": 3521976120,
      "definitions": Array [
        Object {
          "directives": Array [],
          "kind": "OperationDefinition",
          "loc": undefined,
          "name": Object {
            "kind": "Name",
            "loc": undefined,
            "value": "getUser",
          },
          "operation": "query",
          "selectionSet": Object {
            "kind": "SelectionSet",
            "loc": undefined,
            "selections": Array [
              Object {
                "alias": undefined,
                "arguments": Array [
                  Object {
                    "kind": "Argument",
                    "loc": undefined,
                    "name": Object {
                      "kind": "Name",
                      "loc": undefined,
                      "value": "name",
                    },
                    "value": Object {
                      "kind": "Variable",
                      "loc": undefined,
                      "name": Object {
                        "kind": "Name",
                        "loc": undefined,
                        "value": "name",
                      },
                    },
                  },
                ],
                "directives": Array [],
                "kind": "Field",
                "loc": undefined,
                "name": Object {
                  "kind": "Name",
                  "loc": undefined,
                  "value": "user",
                },
                "selectionSet": Object {
                  "kind": "SelectionSet",
                  "loc": undefined,
                  "selections": Array [
                    Object {
                      "alias": undefined,
                      "arguments": Array [],
                      "directives": Array [],
                      "kind": "Field",
                      "loc": undefined,
                      "name": Object {
                        "kind": "Name",
                        "loc": undefined,
                        "value": "id",
                      },
                      "selectionSet": undefined,
                    },
                    Object {
                      "alias": undefined,
                      "arguments": Array [],
                      "directives": Array [],
                      "kind": "Field",
                      "loc": undefined,
                      "name": Object {
                        "kind": "Name",
                        "loc": undefined,
                        "value": "firstName",
                      },
                      "selectionSet": undefined,
                    },
                    Object {
                      "alias": undefined,
                      "arguments": Array [],
                      "directives": Array [],
                      "kind": "Field",
                      "loc": undefined,
                      "name": Object {
                        "kind": "Name",
                        "loc": undefined,
                        "value": "lastName",
                      },
                      "selectionSet": undefined,
                    },
                  ],
                },
              },
            ],
          },
          "variableDefinitions": Array [
            Object {
              "defaultValue": undefined,
              "directives": Array [],
              "kind": "VariableDefinition",
              "loc": undefined,
              "type": Object {
                "kind": "NamedType",
                "loc": undefined,
                "name": Object {
                  "kind": "Name",
                  "loc": undefined,
                  "value": "String",
                },
              },
              "variable": Object {
                "kind": "Variable",
                "loc": undefined,
                "name": Object {
                  "kind": "Name",
                  "loc": undefined,
                  "value": "name",
                },
              },
            },
          ],
        },
      ],
      "kind": "Document",
      "loc": Object {
        "end": 86,
        "source": Object {
          "body": "# getUser
query getUser($name: String) { user(name: $name) { id firstName lastName } }",
          "locationOffset": Object {
            "column": 1,
            "line": 1,
          },
          "name": "gql",
        },
        "start": 0,
      },
    },
    "variables": Object {
      "name": "Clara",
    },
  },
}
`;

exports[`on error returns error data 1`] = `
Object {
  "data": undefined,
  "error": [CombinedError: [Network] ],
  "extensions": undefined,
  "operation": Object {
    "context": Object {
      "fetchOptions": Object {
        "method": "POST",
      },
      "requestPolicy": "cache-first",
      "url": "http://localhost:3000/graphql",
    },
    "key": 2,
    "kind": "query",
    "query": Object {
      "__key": 3521976120,
      "definitions": Array [
        Object {
          "directives": Array [],
          "kind": "OperationDefinition",
          "loc": undefined,
          "name": Object {
            "kind": "Name",
            "loc": undefined,
            "value": "getUser",
          },
          "operation": "query",
          "selectionSet": Object {
            "kind": "SelectionSet",
            "loc": undefined,
            "selections": Array [
              Object {
                "alias": undefined,
                "arguments": Array [
                  Object {
                    "kind": "Argument",
                    "loc": undefined,
                    "name": Object {
                      "kind": "Name",
                      "loc": undefined,
                      "value": "name",
                    },
                    "value": Object {
                      "kind": "Variable",
                      "loc": undefined,
                      "name": Object {
                        "kind": "Name",
                        "loc": undefined,
                        "value": "name",
                      },
                    },
                  },
                ],
                "directives": Array [],
                "kind": "Field",
                "loc": undefined,
                "name": Object {
                  "kind": "Name",
                  "loc": undefined,
                  "value": "user",
                },
                "selectionSet": Object {
                  "kind": "SelectionSet",
                  "loc": undefined,
                  "selections": Array [
                    Object {
                      "alias": undefined,
                      "arguments": Array [],
                      "directives": Array [],
                      "kind": "Field",
                      "loc": undefined,
                      "name": Object {
                        "kind": "Name",
                        "loc": undefined,
                        "value": "id",
                      },
                      "selectionSet": undefined,
                    },
                    Object {
                      "alias": undefined,
                      "arguments": Array [],
                      "directives": Array [],
                      "kind": "Field",
                      "loc": undefined,
                      "name": Object {
                        "kind": "Name",
                        "loc": undefined,
                        "value": "firstName",
                      },
                      "selectionSet": undefined,
                    },
                    Object {
                      "alias": undefined,
                      "arguments": Array [],
                      "directives": Array [],
                      "kind": "Field",
                      "loc": undefined,
                      "name": Object {
                        "kind": "Name",
                        "loc": undefined,
                        "value": "lastName",
                      },
                      "selectionSet": undefined,
                    },
                  ],
                },
              },
            ],
          },
          "variableDefinitions": Array [
            Object {
              "defaultValue": undefined,
              "directives": Array [],
              "kind": "VariableDefinition",
              "loc": undefined,
              "type": Object {
                "kind": "NamedType",
                "loc": undefined,
                "name": Object {
                  "kind": "Name",
                  "loc": undefined,
                  "value": "String",
                },
              },
              "variable": Object {
                "kind": "Variable",
                "loc": undefined,
                "name": Object {
                  "kind": "Name",
                  "loc": undefined,
                  "value": "name",
                },
              },
            },
          ],
        },
      ],
      "kind": "Document",
      "loc": Object {
        "end": 86,
        "source": Object {
          "body": "# getUser
query getUser($name: String) { user(name: $name) { id firstName lastName } }",
          "locationOffset": Object {
            "column": 1,
            "line": 1,
          },
          "name": "gql",
        },
        "start": 0,
      },
    },
    "variables": Object {
      "name": "Clara",
    },
  },
}
`;

exports[`on error returns error data with status 400 and manual redirect mode 1`] = `
Object {
  "data": undefined,
  "error": [CombinedError: [Network] ],
  "extensions": undefined,
  "operation": Object {
    "context": Object {
      "fetchOptions": Object {
        "method": "POST",
      },
      "requestPolicy": "cache-first",
      "url": "http://localhost:3000/graphql",
    },
    "key": 2,
    "kind": "query",
    "query": Object {
      "__key": 3521976120,
      "definitions": Array [
        Object {
          "directives": Array [],
          "kind": "OperationDefinition",
          "loc": undefined,
          "name": Object {
            "kind": "Name",
            "loc": undefined,
            "value": "getUser",
          },
          "operation": "query",
          "selectionSet": Object {
            "kind": "SelectionSet",
            "loc": undefined,
            "selections": Array [
              Object {
                "alias": undefined,
                "arguments": Array [
                  Object {
                    "kind": "Argument",
                    "loc": undefined,
                    "name": Object {
                      "kind": "Name",
                      "loc": undefined,
                      "value": "name",
                    },
                    "value": Object {
                      "kind": "Variable",
                      "loc": undefined,
                      "name": Object {
                        "kind": "Name",
                        "loc": undefined,
                        "value": "name",
                      },
                    },
                  },
                ],
                "directives": Array [],
                "kind": "Field",
                "loc": undefined,
                "name": Object {
                  "kind": "Name",
                  "loc": undefined,
                  "value": "user",
                },
                "selectionSet": Object {
                  "kind": "SelectionSet",
                  "loc": undefined,
                  "selections": Array [
                    Object {
                      "alias": undefined,
                      "arguments": Array [],
                      "directives": Array [],
                      "kind": "Field",
                      "loc": undefined,
                      "name": Object {
                        "kind": "Name",
                        "loc": undefined,
                        "value": "id",
                      },
                      "selectionSet": undefined,
                    },
                    Object {
                      "alias": undefined,
                      "arguments": Array [],
                      "directives": Array [],
                      "kind": "Field",
                      "loc": undefined,
                      "name": Object {
                        "kind": "Name",
                        "loc": undefined,
                        "value": "firstName",
                      },
                      "selectionSet": undefined,
                    },
                    Object {
                      "alias": undefined,
                      "arguments": Array [],
                      "directives": Array [],
                      "kind": "Field",
                      "loc": undefined,
                      "name": Object {
                        "kind": "Name",
                        "loc": undefined,
                        "value": "lastName",
                      },
                      "selectionSet": undefined,
                    },
                  ],
                },
              },
            ],
          },
          "variableDefinitions": Array [
            Object {
              "defaultValue": undefined,
              "directives": Array [],
              "kind": "VariableDefinition",
              "loc": undefined,
              "type": Object {
                "kind": "NamedType",
                "loc": undefined,
                "name": Object {
                  "kind": "Name",
                  "loc": undefined,
                  "value": "String",
                },
              },
              "variable": Object {
                "kind": "Variable",
                "loc": undefined,
                "name": Object {
                  "kind": "Name",
                  "loc": undefined,
                  "value": "name",
                },
              },
            },
          ],
        },
      ],
      "kind": "Document",
      "loc": Object {
        "end": 86,
        "source": Object {
          "body": "# getUser
query getUser($name: String) { user(name: $name) { id firstName lastName } }",
          "locationOffset": Object {
            "column": 1,
            "line": 1,
          },
          "name": "gql",
        },
        "start": 0,
      },
    },
    "variables": Object {
      "name": "Clara",
    },
  },
}
`;

exports[`on success returns response data 1`] = `
Object {
  "data": Object {
    "data": Object {
      "user": 1200,
    },
  },
  "error": undefined,
  "extensions": undefined,
  "hasNext": false,
  "operation": Object {
    "context": Object {
      "fetchOptions": Object {
        "method": "POST",
      },
      "requestPolicy": "cache-first",
      "url": "http://localhost:3000/graphql",
    },
    "key": 2,
    "kind": "query",
    "query": Object {
      "__key": 3521976120,
      "definitions": Array [
        Object {
          "directives": Array [],
          "kind": "OperationDefinition",
          "loc": undefined,
          "name": Object {
            "kind": "Name",
            "loc": undefined,
            "value": "getUser",
          },
          "operation": "query",
          "selectionSet": Object {
            "kind": "SelectionSet",
            "loc": undefined,
            "selections": Array [
              Object {
                "alias": undefined,
                "arguments": Array [
                  Object {
                    "kind": "Argument",
                    "loc": undefined,
                    "name": Object {
                      "kind": "Name",
                      "loc": undefined,
                      "value": "name",
                    },
                    "value": Object {
                      "kind": "Variable",
                      "loc": undefined,
                      "name": Object {
                        "kind": "Name",
                        "loc": undefined,
                        "value": "name",
                      },
                    },
                  },
                ],
                "directives": Array [],
                "kind": "Field",
                "loc": undefined,
                "name": Object {
                  "kind": "Name",
                  "loc": undefined,
                  "value": "user",
                },
                "selectionSet": Object {
                  "kind": "SelectionSet",
                  "loc": undefined,
                  "selections": Array [
                    Object {
                      "alias": undefined,
                      "arguments": Array [],
                      "directives": Array [],
                      "kind": "Field",
                      "loc": undefined,
                      "name": Object {
                        "kind": "Name",
                        "loc": undefined,
                        "value": "id",
                      },
                      "selectionSet": undefined,
                    },
                    Object {
                      "alias": undefined,
                      "arguments": Array [],
                      "directives": Array [],
                      "kind": "Field",
                      "loc": undefined,
                      "name": Object {
                        "kind": "Name",
                        "loc": undefined,
                        "value": "firstName",
                      },
                      "selectionSet": undefined,
                    },
                    Object {
                      "alias": undefined,
                      "arguments": Array [],
                      "directives": Array [],
                      "kind": "Field",
                      "loc": undefined,
                      "name": Object {
                        "kind": "Name",
                        "loc": undefined,
                        "value": "lastName",
                      },
                      "selectionSet": undefined,
                    },
                  ],
                },
              },
            ],
          },
          "variableDefinitions": Array [
            Object {
              "defaultValue": undefined,
              "directives": Array [],
              "kind": "VariableDefinition",
              "loc": undefined,
              "type": Object {
                "kind": "NamedType",
                "loc": undefined,
                "name": Object {
                  "kind": "Name",
                  "loc": undefined,
                  "value": "String",
                },
              },
              "variable": Object {
                "kind": "Variable",
                "loc": undefined,
                "name": Object {
                  "kind": "Name",
                  "loc": undefined,
                  "value": "name",
                },
              },
            },
          ],
        },
      ],
      "kind": "Document",
      "loc": Object {
        "end": 86,
        "source": Object {
          "body": "# getUser
query getUser($name: String) { user(name: $name) { id firstName lastName } }",
          "locationOffset": Object {
            "column": 1,
            "line": 1,
          },
          "name": "gql",
        },
        "start": 0,
      },
    },
    "variables": Object {
      "name": "Clara",
    },
  },
}
`;

exports[`on success uses the mock fetch if given 1`] = `
Object {
  "data": Object {
    "data": Object {
      "user": 1200,
    },
  },
  "error": undefined,
  "extensions": undefined,
  "hasNext": false,
  "operation": Object {
    "context": Object {
      "fetch": [MockFunction] {
        "calls": Array [
          Array [
            "https://test.com/graphql",
            Object {
              "signal": undefined,
            },
          ],
        ],
        "results": Array [
          Object {
            "type": "return",
            "value": Promise {},
          },
        ],
      },
      "fetchOptions": Object {
        "method": "POST",
      },
      "requestPolicy": "cache-first",
      "url": "http://localhost:3000/graphql",
    },
    "key": 2,
    "kind": "query",
    "query": Object {
      "__key": 3521976120,
      "definitions": Array [
        Object {
          "directives": Array [],
          "kind": "OperationDefinition",
          "loc": undefined,
          "name": Object {
            "kind": "Name",
            "loc": undefined,
            "value": "getUser",
          },
          "operation": "query",
          "selectionSet": Object {
            "kind": "SelectionSet",
            "loc": undefined,
            "selections": Array [
              Object {
                "alias": undefined,
                "arguments": Array [
                  Object {
                    "kind": "Argument",
                    "loc": undefined,
                    "name": Object {
                      "kind": "Name",
                      "loc": undefined,
                      "value": "name",
                    },
                    "value": Object {
                      "kind": "Variable",
                      "loc": undefined,
                      "name": Object {
                        "kind": "Name",
                        "loc": undefined,
                        "value": "name",
                      },
                    },
                  },
                ],
                "directives": Array [],
                "kind": "Field",
                "loc": undefined,
                "name": Object {
                  "kind": "Name",
                  "loc": undefined,
                  "value": "user",
                },
                "selectionSet": Object {
                  "kind": "SelectionSet",
                  "loc": undefined,
                  "selections": Array [
                    Object {
                      "alias": undefined,
                      "arguments": Array [],
                      "directives": Array [],
                      "kind": "Field",
                      "loc": undefined,
                      "name": Object {
                        "kind": "Name",
                        "loc": undefined,
                        "value": "id",
                      },
                      "selectionSet": undefined,
                    },
                    Object {
                      "alias": undefined,
                      "arguments": Array [],
                      "directives": Array [],
                      "kind": "Field",
                      "loc": undefined,
                      "name": Object {
                        "kind": "Name",
                        "loc": undefined,
                        "value": "firstName",
                      },
                      "selectionSet": undefined,
                    },
                    Object {
                      "alias": undefined,
                      "arguments": Array [],
                      "directives": Array [],
                      "kind": "Field",
                      "loc": undefined,
                      "name": Object {
                        "kind": "Name",
                        "loc": undefined,
                        "value": "lastName",
                      },
                      "selectionSet": undefined,
                    },
                  ],
                },
              },
            ],
          },
          "variableDefinitions": Array [
            Object {
              "defaultValue": undefined,
              "directives": Array [],
              "kind": "VariableDefinition",
              "loc": undefined,
              "type": Object {
                "kind": "NamedType",
                "loc": undefined,
                "name": Object {
                  "kind": "Name",
                  "loc": undefined,
                  "value": "String",
                },
              },
              "variable": Object {
                "kind": "Variable",
                "loc": undefined,
                "name": Object {
                  "kind": "Name",
                  "loc": undefined,
                  "value": "name",
                },
              },
            },
          ],
        },
      ],
      "kind": "Document",
      "loc": Object {
        "end": 86,
        "source": Object {
          "body": "# getUser
query getUser($name: String) { user(name: $name) { id firstName lastName } }",
          "locationOffset": Object {
            "column": 1,
            "line": 1,
          },
          "name": "gql",
        },
        "start": 0,
      },
    },
    "variables": Object {
      "name": "Clara",
    },
  },
}
`;
